File: c/jnl_ascii_dump.c
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include "jnl.h"

void asciiDump(char *pc, int len);

const int dataBufferSize = 20 * 1024 * 1024;
char *dataBuffer;

void dumpStream(FILE *fp)
{
  while(1)
  {
    struct jnl_header jnlh;
    size_t retFread = fread(&jnlh, sizeof(jnlh), 1, fp);
    if(1 != retFread)
    {
      if(feof(fp))
      {
        break;
      }

      perror("ジャーナルヘッダfread異常");
      exit(1);
    }

    printf("HEADER,");
    printJnlHeader(&jnlh);
    putchar('\n');

    char dataLenBuf[JNL_DATA_LEN + 1];
    memcpy(dataLenBuf, jnlh.dataLen, JNL_DATA_LEN);
    dataLenBuf[JNL_DATA_LEN] = '\0';

    int dataLen = atoi(dataLenBuf);

    retFread = fread(dataBuffer, sizeof(char), dataLen, fp);
    if(retFread < dataLen)
    {
      perror("データfread異常");
      exit(1);
    }

    printf("DATA:");
    asciiDump(dataBuffer, dataLen);
    putchar('\n');
  }
}

void asciiDumpArgs(int argc, char *argv[])
{
  for(int i = 1; i < argc; i++)
  {
    FILE *fp = fopen(argv[i], "r");
    if(NULL == fp)
    {
      perror("fopen異常");
      exit(1);
    }

    dumpStream(fp);

    int retFclose = fclose(fp);
    if(0 != retFclose)
    {
      perror("fclose異常");
      exit(1);
    }
  }
}

int main(int argc, char *argv[])
{
  int isUseStdin = 0;
  if(argc == 1)
  {
    isUseStdin = 1;
  }
  else if(argc == 2 && 0 == strcmp("-h", argv[1]))
  {
    fprintf(stderr, "Usage : %s ファイル名\n", argv[0]);
    exit(1);
  }

  dataBuffer = malloc(dataBufferSize);
  if(NULL == dataBuffer)
  {
    perror("malloc異常");
    exit(1);
  }

  if(isUseStdin)
  {
    dumpStream(stdin);
  }
  else
  {
    asciiDumpArgs(argc, argv);
  }

  return 0;
}

File: python/jnl_ascii_dump.py
#!/usr/bin/env python3

import sys

argv = sys.argv
argc = len(argv)

class jnl_header:
    def __init__(self, header):
        self.year = header[:4]
        self.month = header[4:6]
        self.day = header[6:8]
        self.hour = header[8:10]
        self.minute = header[10:12]
        self.second = header[12:14]
        self.msecond = header[14:17]
        self.kind = header[17:18]
        self.dataLen = header[18:26]


def printHeader(jnlh):
    jnlh_str = "%s/%s/%s,%s:%s:%s.%s,%s,%s" % (jnlh.year,jnlh.month,jnlh.day,jnlh.hour,jnlh.minute,jnlh.second,jnlh.msecond,jnlh.kind,jnlh.dataLen)
    print("HEADER," + jnlh_str)

def printData(data):
    ascii_data = ""
    for byte in data:
        if not byte:
            break
        if (0x20 <= byte and byte <= 0x7e):
            ascii_data += chr(byte)
        else:
            ascii_data += '.'
    print("DATA:" + ascii_data)

def dumpStream():
    stdin = sys.stdin.buffer
    while(1):
        header_size = 26
        header = stdin.read(header_size).decode()
        if not header:
            break
        jnlh = jnl_header(header)
        printHeader(jnlh)
        data_len = jnlh.dataLen
        data = stdin.read(int(data_len))
        printData(data)

def asciiDumpArgs(argc,argv):
    for i in range(1,argc):
        with open(argv[i], "rb") as f:
            while(1):
                header_size = 26
                header = f.read(header_size).decode()
                if not header:
                    break
                jnlh = jnl_header(header)
                printHeader(jnlh)
                data_len = jnlh.dataLen
                data = f.read(int(data_len))
                printData(data)


def main(argc, argv):
    isUseStdin = 0
    if argc== 1:
        isUseStdin = 1
    elif argc == 2 and argv[1] == "-h":
        print("Usage : %s ファイル名\n", argv[0])
        sys.exit()

    if isUseStdin:
        dumpStream()
    else:
        asciiDumpArgs(argc, argv)


#main(argc, argv)

if __name__ == "__main__":
    main(argc, argv)